### PR TITLE
Add check that FSM chaining only involve abstract skills (fix #1085)

### LINKED ISSUE #1085: Check FSM chaining only involves abstract skills
**Is your feature request related to a problem? Please describe.**
As shown in https://github.com/valory-xyz/open-autonomy/issues/1080#issuecomment-1186296993, there is no check whether a chaining of ABCI skills will contain some non-abstract skills. This can cause problems as the non-abstract skill of the ABCI component being chained might interfere with the chaining skill, e.g. by scheduling a concurrent behaviour that makes the main behaviour failing.

**Describe the solution you'd like**
Implement a check in the chaining which assures non of the FSMs which are being chained are concrete.

**Describe alternatives you've considered**
n/a

**Additional context**
n/a

### PR DESCRIPTION
**Pull request by marcofavorito**
## Proposed changes

This PR implements a check in the `packages/valory/skills/abstract_round_abci/abci_app_chain.py::chain` function that requires operands to be _abstract_ AbciApp. This is done by adding a private class attribute `_is_abstract: bool`, which defaults to `True`, but it is set from the `SharedState` of the ABCI skill that implements it.

This is useful to prevent a bad omission caused by `registration_abci` skill not being abstract, as shown in #1080 .

## Fixes

Fix #1085 

## Types of changes

What types of changes does your code introduce? (A **breaking change** is a fix or feature that would cause existing functionality and APIs to not work as expected.)
_Put an `x` in the box that applies_

- [X] Non-breaking fix (non-breaking change which fixes an issue)
- [ ] Breaking fix (breaking change which fixes an issue)
- [ ] Non-breaking feature (non-breaking change which adds functionality)
- [ ] Breaking feature (breaking change which adds functionality)
- [ ] Refactor (non-breaking change which changes implementation)
- [ ] Messy (mixture of the above - requires explanation!)

## Checklist

_Put an `x` in the boxes that apply._

- [X] I have read the [CONTRIBUTING](../blob/main/CONTRIBUTING.md) doc
- [X] I am making a pull request against the `main` branch (left side). Also you should start your branch off our `main`.
- [X] Lint and unit tests pass locally with my changes
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have locally run services that could be impacted and they do not present failures derived from my changes

## Further comments

n.a

### CODE REVIEW

**Review by DavidMinarsch — APPROVED** [review]

File: tests/test_skills/base.py [hunk 1]
```diff
@@ -376,3 +377,14 @@ def teardown(cls) -> None:
...
+
+class DummyContext:
+    """Dummy Context class for testing shared state initialization."""
+
+    class params:
+        """Dummy param variable."""
+
+        round_timeout_seconds: float = 1.0
+
+    _skill: MagicMock = MagicMock()
```
Comment (Adamantios) [inline tests/test_skills/base.py:390]: :+1:

File: packages/valory/skills/abstract_round_abci/models.py [hunk 2]
```diff
@@ -101,11 +101,20 @@ def _ensure(cls, key: str, kwargs: Dict) -> Any:
...
+        *args: Any,
+        abci_app_cls: Type[AbciApp],
+        skill_context: SkillContext,
+        **kwargs: Any,
+    ) -> None:
         """Initialize the state."""
         self.abci_app_cls = self._process_abci_app_cls(abci_app_cls)
+        self.abci_app_cls._is_abstract = (
+            skill_context._skill.configuration.is_abstract_component  # pylint: disable=protected-access
+        )
```
Comment (Adamantios) [inline packages/valory/skills/abstract_round_abci/models.py:115]: @angrybayblade I'd suggest that we create a property for `is_abstract_component` at `open-aea`'s `SkillContext`.

**Review by Adamantios — APPROVED** [review]
LGTM, only some minor comments.

Also, let's first see if e2e tests still timeout after merging `test/registration` on this branch.

File: packages/valory/skills/abstract_round_abci/models.py (hunk 2 above)
Comment (marcofavorito) [reply packages/valory/skills/abstract_round_abci/models.py:115]: the problem is that `is_abstract` should not be modified arbitrarily, but only propertly by `SharedState`, hence it was a protected class attributes.

We could have a public access method though.

Comment (Adamantios) [reply packages/valory/skills/abstract_round_abci/models.py:115]: #1093.

**Pull request merged**